src/Featurizers/PCAFeaturizer.h

```diff
@@ -0,0 +1,267 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include "../Featurizer.h"
+#include "../Archive.h"
+
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         PCATransformer
+///  \brief         Contains PCAComponents and use PCAComponents to project
+///                 matrix for dimensionality reduction, also provides
+///                 PCAComponents retriving
+///
+template <typename MatrixT>
+class PCATransformer : public StandardTransformer<MatrixT, MatrixT> {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType                                   = StandardTransformer<MatrixT, MatrixT>;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    PCATransformer(MatrixT eigenvalues, MatrixT eigenvectors);
+    PCATransformer(Archive &ar);
+
+    ~PCATransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(PCATransformer);
+
+    bool operator==(PCATransformer const &other) const;
+
+    void save(Archive &ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    MatrixT const                                    _eigenvalues;
+    MatrixT const                                    _eigenvectors;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Methods
+    // |
+    // ----------------------------------------------------------------------
+     // MSVC has problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        if (input.size() == 0)
+            throw std::invalid_argument("Input matrix size() invalid");
+
+        if (input.cols() != _eigenvectors.cols())
```

Comment:
>if (input.cols() != _eigenvectors.cols()) [](start = 8, length = 41)

Can you add a comment if this code is correct? It seems strange to me that it expects the col values to match.

Reply:
because this is simplified(mathematically) from input.col() == _eigenvectors.transpose().rows(). I 'll add comments for it